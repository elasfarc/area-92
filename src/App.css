.App {
  text-align: center;
}

@media (prefers-reduced-motion: no-preference) {
  .App-logo {
    animation: App-logo-spin infinite 20s linear;
  }
}

.App-header {
  background-color: #282c34;
  min-height: 100vh;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  font-size: calc(10px + 2vmin);
  color: white;
}

.App-link {
  color: #61dafb;
}

@keyframes App-logo-spin {
  from {
    transform: rotate(0deg);
  }
  to {
    transform: rotate(360deg);
  }
}

/* app styles  */
@import url("https://fonts.googleapis.com/css2?family=Playfair+Display:ital,wght@0,400;0,500;0,600;0,700;0,800;0,900;1,400;1,500;1,600;1,700;1,800;1,900&family=Roboto:ital,wght@0,100;0,300;0,400;0,500;0,700;0,900;1,100;1,300;1,400;1,500;1,700;1,900&display=swap");
* {
  padding: 0;
  margin: 0;
  box-sizing: border-box;
}

body {
  font-family: "Roboto", sans-serif;
}
img {
  max-width: 100%;
  object-fit: cover;
}

.flex {
  display: flex;
}
.space-between {
  justify-content: space-between;
}

.cross-center {
  align-items: center;
}
.gap-1 {
  gap: 1rem;
}
.gap-2 {
  gap: 2rem;
}
.gap-3 {
  gap: 3rem;
}
.reverse-order {
  order: -1;
}
.text-center {
  text-align: center;
}
.mb-1 {
  margin-bottom: 0.125rem;
}
.container {
  width: 90%;
  max-width: 700px;
  margin: 0 auto;
}
.heading {
  font-family: "Playfair Display", serif;
}
.home {
  max-width: 700px;
}
.box {
  padding: 4em 0;
  cursor: pointer;
  color: white;
  background-color: #ff4f92;
}

.box__heading {
  font-size: 3rem;
  margin-bottom: 1rem;
  text-transform: uppercase;
  letter-spacing: 2px;
}
.box__img {
  opacity: 50%;
  width: 30%;
}
