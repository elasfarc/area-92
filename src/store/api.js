// ACTION TYPES
export const API_CALL_REQUEST = "/api/request";
export const API_CALL_SUCCESS = "/api/success";
export const API_CALL_FAIL = "/api/fail";

// ACTION CREATOR
export const requestApiCall = (payload) => ({
  type: API_CALL_REQUEST,
  payload,
});
export const onApiSuccess = () => ({ type: API_CALL_SUCCESS });
export const onApiFail = () => ({ type: API_CALL_FAIL });
