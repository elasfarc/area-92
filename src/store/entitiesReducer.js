import { combineReducers } from 'redux'
import countryReducer from './country'

export default  combineReducers({
  countries: countryReducer
})