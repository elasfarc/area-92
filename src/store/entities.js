import * as apiActions from "./api";

// ACTION TYPES
const COUNTRIES_REQUESTED = "/countries/request";
const COUNTRIES_LOADED = "/countries/load";
const STATES_LOADED = "/country/states";
const CITIES_TRANSFORM_TO_GEO = "country/cities/transform2Geo";
const CITY_WEATHER_LOADED = "/weather/city";

// async thunk
export const loadCountriesData = () => async (dispatch) => {
  dispatch({ type: COUNTRIES_REQUESTED });
  try {
    const { data: countries } = await (
      await fetch(
        `https://countriesnow.space/api/v0.1/countries/info?returns=flag,currency,cities`
      )
    ).json();
    const { data: capitals } = await (
      await fetch(`https://countriesnow.space/api/v0.1/countries/capital`)
    ).json();
    dispatch(apiActions.onApiSuccess());
    dispatch({
      type: COUNTRIES_LOADED,
      payload: { countries, capitals },
    });
  } catch (error) {
    dispatch(apiActions.onApiFail());
  }
};

// ACTION CREATOR

export const getCountryStates = (country) =>
  apiActions.requestApiCall({
    url: `https://countriesnow.space/api/v0.1/countries/states`,
    method: "POST",
    body: { country },
    onStart: COUNTRIES_REQUESTED,
    onSuccess: STATES_LOADED,
    info: country,
  });

export const transformCitiesToGeo = ({ country, cities }) =>
  apiActions.requestApiCall({
    url: `http://open.mapquestapi.com/geocoding/v1/batch?key=4qiFlGPdSaIgG97k8C3AyJRKLBpq1jtJ`,
    method: "POST",
    body: { locations: cities.map((city) => ({ city, country })) },
    onStart: COUNTRIES_REQUESTED,
    onSuccess: CITIES_TRANSFORM_TO_GEO,
    info: country,
  });

export const getGeoWeather = () => ({});

// REDUCER
const initialState = {
  countries: [],
  isLoading: false,
};

const reducer = (state = initialState, { type, payload }) => {
  if (type === COUNTRIES_REQUESTED) return { ...state, isLoading: true };

  if (type === COUNTRIES_LOADED) {
    const { countries, capitals } = payload;
    return {
      ...state,
      isLoading: false,
      countries: countries.map((country) => ({
        ...country,
        ...capitals.find(({ capital, name }) => country.name === name),
      })),
    };
  }

  if (type === STATES_LOADED) {
    const {
      response: {
        data: { states },
      },
      info: countryName,
    } = payload;

    return {
      ...state,
      isLoading: false,
      countries: state.countries.map((country) =>
        country.name === countryName ? { ...country, states } : country
      ),
    };
  }

  if (type === CITIES_TRANSFORM_TO_GEO) {
    return {
      ...state,
      isLoading: false,
      countries: state.countries.map((country) =>
        country.name === payload.info
          ? {
              ...country,
              cities: payload.response.results.map(
                ({
                  providedLocation,
                  locations: [{ latLng, mapUrl, adminArea3 }],
                }) => ({
                  name: providedLocation.city,
                  country: country.name,
                  state: adminArea3,
                  latLng,
                  mapUrl,
                })
              ),
            }
          : country
      ),
    };
  }
  if (type === CITY_WEATHER_LOADED) return state;
  return state;
};

export default reducer;
