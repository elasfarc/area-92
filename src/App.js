import "./App.css";
import {
  BrowserRouter as Router,
  Switch,
  Route,
  Link,
  useParams,
  useRouteMatch,
} from "react-router-dom";
import * as entitiesActions from "./store/entities";
import Home from "./components/Home";
import Country from "./components/Country";

function App() {
  return (
    <Router>
      <Route path="/" exact>
        <Home />
      </Route>
      <Route path="/:country" exact>
        <Country />
      </Route>
    </Router>
  );
}

export default App;
