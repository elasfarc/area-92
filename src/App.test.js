import renderer from "react-test-renderer";
import { Provider } from "react-redux";
import App from "./App";
import store from "./store/configureStore";

const MockApp = () => (
  <Provider store={store}>
    <App />
  </Provider>
);

describe("Mission component", () => {
  it("should match the snapshot", () => {
    const tree = renderer.create(<MockApp />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
