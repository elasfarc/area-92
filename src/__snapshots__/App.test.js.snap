// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Mission component should match the snapshot 1`] = `
Array [
  <div
    className="navbar mb-1"
  >
    <div
      className="container  flex"
    >
      <button
        onClick={[Function]}
        type="button"
      >
        <svg
          aria-hidden="true"
          className="svg-inline--fa fa-angle-left fa-w-8 "
          data-icon="angle-left"
          data-prefix="fas"
          focusable="false"
          role="img"
          style={Object {}}
          viewBox="0 0 256 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M31.7 239l136-136c9.4-9.4 24.6-9.4 33.9 0l22.6 22.6c9.4 9.4 9.4 24.6 0 33.9L127.9 256l96.4 96.4c9.4 9.4 9.4 24.6 0 33.9L201.7 409c-9.4 9.4-24.6 9.4-33.9 0l-136-136c-9.5-9.4-9.5-24.6-.1-34z"
            fill="currentColor"
            style={Object {}}
          />
        </svg>
      </button>
      <form
        onSubmit={[Function]}
      >
        <input
          className="nav__search"
          name="inputSearch"
          onChange={[Function]}
        />
        <button
          type="submit"
        >
          GO
        </button>
      </form>
    </div>
  </div>,
  <h1
    className="text-center"
  >
    Loading Countries...
  </h1>,
]
`;
